### PR TITLE
Basic rewriting algorithm/tree walker

### PR DESCRIPTION
**Pull request by Kieranoski702**
As per the rust meeting today, this is the basic form of the tree walker and rewriting algorithm. Currently, this is just a depth-first search that tries to apply all the rules to each expression in the tree - Returning the new expression if it can be rewritten or recursing through the expression's sub-expressions if it cannot.

Very soon I will be optimizing this slightly by marking nodes as dirty or clean and only searching through dirty nodes which will save some work. 

Everything seems to build properly on fedora Linux but there are most likely some errors in here that I have not yet spotted. If you see any errors or spots for improvement please let me know!

### CODE REVIEW

File: crates/conjure_core/src/ast.rs [hunk 1]
```diff
@@ -102,3 +102,71 @@ pub enum Expression {
     SumLeq(Vec<Expression>, Box<Expression>),
     Ineq(Box<Expression>, Box<Expression>, Box<Expression>),
 }
+
+impl Expression {
```
Comment (ozgurakgun) [inline crates/conjure_core/src/ast.rs:106]: this is the kind of thing a lens/optics library can automate I am guessing (without looking into the rust libs)

**Comment by ozgurakgun** [comment]
LGTM, can you do a quick review please @lixitrixi and @niklasdewally

File: crates/conjure_core/src/ast.rs (hunk 1 above)
Comment (niklasdewally) [reply crates/conjure_core/src/ast.rs:106]: Shouldn't stop merging this now, but definitely would make this more scalable.

https://rust-unofficial.github.io/patterns/functional/optics.html
https://github.com/TOETOE55/lens-rs/blob/master/guide.md

Comment (niklasdewally) [reply crates/conjure_core/src/ast.rs:106]: `lens-rs` seems quite nice:

```
#[derive(Copy, Clone, Debug, Review, Prism)]
enum Either<L, R> {
    #[optic]
    Left(L), // generate optics::Left
    #[optic]
    Right(R), // generate optics::Right
}

#[derive(Copy, Clone, Debug, Lens)]
struct Tuple<A, B>(#[optic] A, #[optic] B); 
// use optics::_0 or optics::_1 to access it

#[derive(Copy, Clone, Debug, Lens)]
struct Foo<A, B> {
    #[optic]
    a: A, // generate optics::a
    #[optic]
    b: B, // generate optics::b
}

#[derive(Clone, Debug, Lens)]
struct Bar {
    #[optic]
    a: String, // generate optics::a, same as above
    #[optic]
    c: i32,    // generate optics::c
}
```

Comment (niklasdewally) [reply crates/conjure_core/src/ast.rs:106]: Querying , modification, construction works through a macro:

```rs
let y: _ = optic!(Ok.Some.Err).review((1, 2));
// Ok(Some(Err((1, 2))))
```

Comment (niklasdewally) [reply crates/conjure_core/src/ast.rs:106]: If the exact syntax isn't too our liking, we could construct something similar with macros without too much effort.

File: conjure_oxide/src/lib.rs [hunk 2]
```diff
@@ -1,7 +1,8 @@
 pub mod error;
 pub mod find_conjure;
 pub mod parse;
-mod rules;
+mod rewrite;
+pub mod rules;
```
Comment (niklasdewally) [inline conjure_oxide/src/lib.rs:5]:
```suggestion
mod rewrite;
mod rules;
```

**Review by niklasdewally — COMMENTED** [review]
LGTM, but I think it might be good to have a test case for this to see that it works (cargo currently says that its dead code for me).

Possibly just copy some of Felix's ones from #146 (which go from a model defined in Rust to a solution in minion), and replace the manual rule applications with a call to rewrite?

**Review by lixitrixi — CHANGES_REQUESTED** [review]
Apologies for the late review! The code itself looks good to me, but now that there are enough rules to test this running (after a rebase) could we add a test with XYZ to Minion? I know @niklasdewally already mentioned this before, it could just be a copy of the manual test, with a call to `rewrite` instead.

**Review by lixitrixi — APPROVED** [review]

**Review by niklasdewally — APPROVED** [review]

**Comment by Kieranoski702** [comment]
This should be all ready to merge now. I'll get a new PR with the tests up at some point this week (hopefully before the meeting on Wednesday). After that, I think I'll take a look at using the lens library Nik mentioned here and update this to use the lenses instead. Let me know if this PR needs anything else though!

**Pull request merged**